extensions/amp-date-picker/0.1/amp-date-picker.js

```diff
@@ -0,0 +1,612 @@
+    this.blocked_ = new DatesList(
+        blocked ? blocked.split(DATE_SEPARATOR) : []);
+
+    const highlighted = this.element.getAttribute('highlighted');
+    /** @private @const */
+    this.highlighted_ = new DatesList(
+        highlighted ? highlighted.split(DATE_SEPARATOR) : []);
+
+    /** @private @const */
+    this.container_ = this.element.ownerDocument.createElement('div');
+
+    const type = this.element.getAttribute('type') || 'single';
+    /** @private @const */
+    this.picker_ = (type === 'range' ?
+        createDateRangePicker() :
+        createSingleDatePicker());
+
+    /** @private @const */
+    this.props_ = this.getProps_();
+
+    /** @private @const */
+    this.elementTemplates_ = this.parseElementTemplates_();
+
+    /** @private {!Array<!DateTemplateMapDef>} */
+    this.srcTemplates_ = [];
+
+    /** @private {?Element} */
+    this.srcDefaultTemplate_ = null;
+
+    /** @private @const */
+    this.isRTL_ = isRTL(/** @type {!Document} */ (this.ampdoc_.getRootNode()));
+
+    /** @private {?function(?):?} */
+    this.templateThen_ = null;
+
+    /** @private @const */
+    this.registerAction_ = this.registerAction.bind(this);
+
+    const locale = this.element.getAttribute('locale') || DEFAULT_LOCALE;
+    this.moment_.locale(locale);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @override */
+  buildCallback() {
+    user().assert(isExperimentOn(this.win, TAG),
+        `Experiment ${TAG} is disabled.`);
+
+    this.element.appendChild(this.container_);
+    this.render();
+    this.element.setAttribute('i-amphtml-date-picker-attached', '');
+  }
+
+  /** @override */
+  layoutCallback() {
+    return this.parseSrcTemplates_(this.fetchSrcTemplates_());
+  }
+
+  /**
+   * Fetch the JSON from the URL specified in the src attribute.
+   * @return {?Promise<!JsonObject|!Array<JsonObject>>}
+   */
+  fetchSrcTemplates_() {
+    if (this.element.getAttribute('src')) {
+      return fetchBatchedJsonFor(this.ampdoc_, this.element);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Create an array of objects mapping dates to templates.
+   * @param {?Promise<!JsonObject|!Array<JsonObject>>} srcTemplatePromise
+   */
+  parseSrcTemplates_(srcTemplatePromise) {
+    if (!srcTemplatePromise) {
+      return Promise.resolve();
+    }
+
+    return srcTemplatePromise.then(srcJson => {
+      const templates = srcJson && srcJson['templates'];
+      if (!templates) {
+        return;
+      }
+
+      const srcTemplates = templates
+          .filter(t => t.dates)
+          .map(t => ({
+            dates: new DatesList(t.dates),
+            template: this.ampdoc_.getRootNode().querySelector(
+                `#${t.id}[date-template]`),
+          }));
+      this.srcTemplates_ = srcTemplates;
+
+      const defaultTemplate = templates
+          .filter(t => t.dates == null)
+          .map(t => this.ampdoc_.getElementById(t.id))[0];
```

Comment:
ditto for `getElementById`, scoping to this element via querySelector.

(although ids have to be unique, we don't restrict it and I have seen bugs in AMP pages caused by duplicated ids, this sort of scoping can help be more forgiving of such user mistakes)